src/Plugin/authentication/OAuth2ServerAuthentication.php

```diff
@@ -0,0 +1,74 @@
+<?php
+
+namespace Drupal\restful\Plugin\authentication;
+
+use Drupal\Component\Plugin\PluginBase;
+use Drupal\restful\Exception\UnauthorizedException;
+use Drupal\restful\Http\RequestInterface;
+use Drupal\restful\Plugin\ResourcePluginManager;
+
+/**
+ * Authentication support for oauth2_server.
+ *
+ * @Authentication(
+ *   id = "oauth2_auth",
+ *   label = "OAuth2 authentication",
+ *   description = "Authenticate requests based on oauth2_server auth.",
+ * )
+ */
+class OAuth2ServerAuthentication extends Authentication {
+
+  /**
+   * {@inheritdoc}
+   */
+  public function applies(RequestInterface $request) {
+    return module_exists('oauth2_server') && $this->getResourcePluginIdFromRequest();
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function authenticate(RequestInterface $request) {
+    $oauth2_info = $this->getOAuth2Info($request);
+    if (!$oauth2_info) {
+      return NULL;
+    }
+
+    $result = oauth2_server_check_access($oauth2_info['server'], $oauth2_info['scope']);
+    if ($result instanceof \OAuth2\Response) {
+      throw new UnauthorizedException($result->getResponseBody(), $result->getStatusCode());
+    }
+    elseif (is_array($result) && !empty($result['user_id'])) {
+      return user_load($result['user_id']);
+    }
+  }
+
+//  protected function getOAuth2Info() {
+//    return [variable_get('oauth2_server_restful_server'), variable_get('oauth2_server_restful_scope')];
+//  }
+
+  protected function getOAuth2Info($request) {
+    $plugin_id = $this->getResourcePluginIdFromRequest();
+    $plugin = ResourcePluginManager::create('cache', $request)->getDefinition($plugin_id);
+
+    $server = !empty($plugin['oauth2Server']) ? $plugin['oauth2Server'] : variable_get('oauth2_server_restful_server');
```

Comment:
The `options` key is on `@Authentication` annotation. I must put that on the OAuth2Authentcation class. That does not help the use case where resources would want to specify the server and scope.